routing/router.go

```diff
@@ -2293,3 +2293,278 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+	fromNode route.Vertex
+}
+
+func (e ErrNoChannel) Error() string {
+	return fmt.Sprintf("no matching outgoing channel available for "+
+		"node %v (%v)", e.position, e.fromNode)
+}
+
+// BuildRoute returns a fully specified route based on a list of pubkeys. If
+// amount is nil, the minimum routable amount is used. To force a specific
+// outgoing channel, use the outgoingChan parameter.
+func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
+	hops []route.Vertex, outgoingChan *uint64,
+	finalCltvDelta int32) (*route.Route, error) {
+
+	log.Debugf("BuildRoute called: hopsCount=%v, amt=%v",
+		len(hops), amt)
+
+	// We'll attempt to obtain a set of bandwidth hints that helps us select
+	// the best outgoing channel to use in case no outgoing channel is set.
+	bandwidthHints, err := generateBandwidthHints(
+		r.selfNode, r.cfg.QueryBandwidth,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Allocate a list that will contain the selected channels for this
+	// route.
+	edges := make([]*channeldb.ChannelEdgePolicy, len(hops))
+
+	// Keep a running amount and the maximum for this route. Leave amt set
+	// to its default value of zero for minimum amount route builds.
+	amts := runningAmounts{
+		max: lnwire.MilliSatoshi(^uint64(0)),
+	}
+	if amt != nil {
+		amts.amt = *amt
+	}
+
+	// Traverse hops backwards to accumulate fees in the running amounts.
+	source := r.selfNode.PubKeyBytes
+	for i := len(hops) - 1; i >= 0; i-- {
+		toNode := hops[i]
+
+		var fromNode route.Vertex
+		if i == 0 {
+			fromNode = source
+		} else {
+			fromNode = hops[i-1]
+		}
+
+		localChan := i == 0
+
+		// Iterate over candidate channels to select the channel
+		// to use for the final route.
+		var (
+			bestEdge      *channeldb.ChannelEdgePolicy
+			bestAmts      *runningAmounts
+			bestBandwidth lnwire.MilliSatoshi
+		)
+
+		cb := func(tx *bbolt.Tx,
+			edgeInfo *channeldb.ChannelEdgeInfo,
+			_, inEdge *channeldb.ChannelEdgePolicy) error {
+
+			chanID := edgeInfo.ChannelID
+
+			// Apply outgoing channel restriction is active.
+			if localChan && outgoingChan != nil &&
+				chanID != *outgoingChan {
+
+				return nil
+			}
+
+			// No unknown policy channels.
+			if inEdge == nil {
+				return nil
+			}
+
+			// Before we can process the edge, we'll need to
+			// fetch the node on the _other_ end of this
+			// channel as we may later need to iterate over
+			// the incoming edges of this node if we explore
+			// it further.
+			chanFromNode, err := edgeInfo.FetchOtherNode(
+				tx, toNode[:],
+			)
+			if err != nil {
+				return err
+			}
+
+			// Continue searching if this channel doesn't
+			// connect with the previous hop.
+			if chanFromNode.PubKeyBytes != fromNode {
+				return nil
+			}
+
+			// Validate whether this channel's policy is satisfied
+			// and obtain the new running amounts if this channel
+			// was to be selected.
+			newAmts, err := r.checkOrUpdateAmt(
+				inEdge, edgeInfo.Capacity, amts, localChan,
+				amt == nil,
```

Comment:
could maybe extract this into `bool` to make it clear what it is

Reply:
extracted